src/internet_identity/src/archive.rs

```diff
@@ -0,0 +1,179 @@
+use CanisterInstallMode::Upgrade;
+
+pub async fn deploy_archive(wasm: ByteBuf) -> DeployArchiveResult {
+    let archive_state = state::persistent_state_mut(|persistent_state| {
+        let archive_state = persistent_state.archive_info.state.clone();
+        if archive_state == ArchiveState::NotCreated {
+            // lock archive creation because of async operation
+            persistent_state.archive_info.state = ArchiveState::CreationInProgress;
+        }
+        archive_state
+    });
+
+    match archive_state {
+        ArchiveState::NotCreated => create_and_install_archive(wasm.into_vec()).await,
+        ArchiveState::CreationInProgress => DeployArchiveResult::CreationInProgress,
+        ArchiveState::Created(data) => {
+            match install_archive(data.archive_canister, wasm.into_vec()).await {
+                Ok(()) => DeployArchiveResult::Success(data.archive_canister),
+                Err(err) => UpgradeFailed(err),
+            }
+        }
+    }
+}
+
+async fn create_and_install_archive(wasm: Vec<u8>) -> DeployArchiveResult {
+    match create_archive().await {
+        Err(e) => {
+            state::persistent_state_mut(|persistent_state| {
+                // unlock archive creation again
+                persistent_state.archive_info.state = ArchiveState::NotCreated
+            });
+            return CreationFailed(format!("failed to create archive: {:?}", e));
+        }
+        Ok(data) => {
+            let archive_canister_id = data.archive_canister;
+            state::persistent_state_mut(|persistent_state| {
+                // safe archive info permanently
+                persistent_state.archive_info.state = ArchiveState::Created(data)
+            });
+            match install_archive(archive_canister_id, wasm).await {
+                Ok(()) => DeployArchiveResult::Success(archive_canister_id),
+                Err(err) => UpgradeFailed(err),
+            }
+        }
+    }
+}
+
+async fn create_archive() -> CallResult<ArchiveData> {
+    let (result,) = create_canister(CreateCanisterArgument { settings: None }).await?;
+    Ok(ArchiveData {
+        sequence_number: 0,
+        archive_canister: result.canister_id,
+    })
+}
+
+async fn install_archive(archive_canister: Principal, wasm_module: Vec<u8>) -> Result<(), String> {
+    let expected_hash = verify_wasm_hash(&wasm_module)?;
+
+    let (archive_status,) = canister_status(CanisterIdRecord {
+        canister_id: archive_canister,
+    })
+    .await
+    .map_err(|err| format!("failed to retrieve archive status: {:?}", err))?;
+
+    let module_hash = archive_status.module_hash;
+    if module_hash
+        .clone()
+        .map(|hash| hash == expected_hash.to_vec())
+        .unwrap_or(false)
+    {
+        // Don't do anything further if the archive canister has the given module already installed
+        return Ok(());
+    }
+
+    let mode = match module_hash {
+        None => Install,
+        Some(_) => Upgrade,
+    };
```

Comment:
As far as I can tell this creates an `Option<bool>` with the following semantics (I've renamed `module_hash` to `current_hash` and `expected_hash` to `new_hash` for clarity):

* if `current_hash` is not set, return `None` -> Install
* if `current_hash` is the same as `new_hash`, then return `Some(true)` -> Upgrade
* if `current_hash` is _not_ the same as `new_hash`, then return `Some(false)` -> Upgrade


Couple questions:
* Why even upgrade if the hash is the same? is that in case of new arguments?
* Do you think explicitly matching on `current_hash` (`module_hash`) might be simpler to read/understand?

Reply:
The `if` clause above checks the hash and exits early if `current_hash` is the same as `new_hash`. But I agree, I will reshuffle things to make it more clear. 👍